rlu/src/stm/mod.rs

```diff
@@ -0,0 +1,659 @@
+        TVar {
+            original: Arc::new(Pair::new(val)),
+            lock: VersionLock::new(self.global.version()),
+        }
+    }
+}
+
+#[cfg(feature = "threaded")]
+impl<T> Transaction<T>
+where
+    T: Clone + Debug,
+{
+    pub fn new(version: usize, id: usize) -> Self {
+        Self {
+            version,
+            read: HashSet::new(),
+            write: HashMap::new(),
+            id,
+        }
+    }
+
+    /// This function loads the value from the transactional variable ([`TVar`]) and checks
+    /// for version consistency. If the value is present in a write set, this to-be-written value
+    /// will be returned. In case there is a version mismatch, or the transactional variable is
+    /// locked, an error will be returned and the [`Transaction`] will be retried.
+    pub fn load(&mut self, tvar: &TVar<T>) -> Result<T, TxError> {
+        self.read.insert(tvar.clone());
+
+        if self.write.contains_key(tvar) {
+            return Ok(self.write.get(tvar).unwrap().clone());
+        }
+
+        if tvar.lock.is_locked() {
+            match tvar.is_locked_by(self.id) {
+                ThreadLockState::Same | ThreadLockState::None => {
+                    tvar.lock.unlock()?;
+                }
+                ThreadLockState::Foreign => {
+                    info!(
+                        "LOAD({:?}): PRECHECK TRANSACTION LOCKED BY ({:?})",
+                        self.id,
+                        tvar.locked_by()
+                    );
+                    return Err(TxError::TransactionLocked);
+                }
+            }
+        }
+
+        let pre_version = tvar.lock.version();
+
+        let data = tvar.original.value.lock().map_err(|e| TxError::LockPresent)?;
+        info!("LOAD({:?}): TVAR CONTENTS ({:?})", self.id, data);
+        let post_version = tvar.lock.version();
+
+        let is_locked = tvar.lock.is_locked();
+        let version_mismatch = pre_version != post_version;
+        let stale_object = pre_version > self.version;
+
+        match is_locked || version_mismatch || stale_object {
+            true => {
+                info!(
+                    "LOAD({:?}): VERSION MISMATCH ({}), STALE OBJECT ({}), PRE_VERSION ({}), TRANSACTION_VERSION ({})",
+                    self.id, version_mismatch, stale_object, pre_version, self.version
+                );
+                Err(TxError::VersionMismatch)
+            }
+            false => Ok((*data).clone()),
+        }
+    }
+
+    /// this writes the value into the transactional log
+    pub fn store(&mut self, tvar: &TVar<T>, value: T) -> Result<(), TxError> {
+        self.write.insert(tvar.clone(), value);
+
+        Ok(())
+    }
+
+    #[inline(always)]
+    fn lock_write_set(&self) -> Result<(), TxError> {
+        let fail_lock_write = AtomicBool::new(false);
+
+        for tvar in self.write.keys() {
+            info!("TRANSACTION({:?}): WRITE LOCK", self.id);
+
+            if tvar.lock.try_lock().is_err() {
```

Comment:
The current implementation only returns `Ok(_)`

Reply:
This relates to your comment above, since try-lock is unbounded, there is no way to return an `Err`, but `Ok`.